File: .github/scripts/notion-slack-sync.js
const { Client } = require('@notionhq/client');
const { WebClient } = require('@slack/web-api');
const fs = require('fs');
const path = require('path');

// Initialize clients
const notion = new Client({ auth: process.env.NOTION_TOKEN });
const slack = new WebClient(process.env.SLACK_BOT_TOKEN);

const ROOT_PAGE_ID = process.env.NOTION_ROOT_PAGE_ID;
const SLACK_CHANNEL = process.env.SLACK_CHANNEL_ID;
const CACHE_DIR = path.join(__dirname, '..', 'cache');
const CACHE_FILE = path.join(CACHE_DIR, 'posted-pages.json');

// Ensure cache directory exists
if (!fs.existsSync(CACHE_DIR)) {
  fs.mkdirSync(CACHE_DIR, { recursive: true });
}

// Load cache of already posted pages
function loadCache() {
  try {
    if (fs.existsSync(CACHE_FILE)) {
      const data = fs.readFileSync(CACHE_FILE, 'utf8');
      return JSON.parse(data);
    }
  } catch (error) {
    console.error('Error loading cache:', error);
  }
  return { postedPages: [] };
}

// Save cache
function saveCache(cache) {
  try {
    fs.writeFileSync(CACHE_FILE, JSON.stringify(cache, null, 2));
  } catch (error) {
    console.error('Error saving cache:', error);
  }
}

// Get all pages recursively from a root page
async function getAllPagesRecursive(pageId, pages = [], isRoot = true) {
  try {
    // Get the page details (skip for root page as we don't want to include it)
    if (!isRoot) {
      const page = await notion.pages.retrieve({ page_id: pageId });
      if (page.object === 'page') {
        pages.push(page);
      }
    }
    
    // Get all child blocks
    let hasMore = true;
    let cursor = undefined;
    
    while (hasMore) {
      const response = await notion.blocks.children.list({
        block_id: pageId,
        page_size: 100,
        start_cursor: cursor
      });
      
      // Process each block
      for (const block of response.results) {
        if (block.type === 'child_page') {
          // Recursively get this child page and its children
          await getAllPagesRecursive(block.id, pages, false);
        }
      }
      
      hasMore = response.has_more;
      cursor = response.next_cursor;
    }
  } catch (error) {
    console.error(`Error fetching pages for ${pageId}:`, error.message);
  }
  
  return pages;
}

// Extract page title
function getPageTitle(page) {
  try {
    if (page.properties) {
      // Try different property types
      const titleProp = Object.values(page.properties).find(
        prop => prop.type === 'title'
      );
      
      if (titleProp && titleProp.title && titleProp.title.length > 0) {
        return titleProp.title.map(t => t.plain_text).join('');
      }
    }
    
    // Fallback to child_page title if available
    if (page.child_page && page.child_page.title) {
      return page.child_page.title;
    }
    
    return 'Untitled';
  } catch (error) {
    console.error('Error extracting title:', error);
    return 'Untitled';
  }
}

// Get author from page
function getAuthor(page) {
  try {
    // Try to find author in properties
    if (page.properties) {
      const authorProp = Object.values(page.properties).find(
        prop => prop.type === 'people' || 
               prop.type === 'created_by' ||
               (prop.type === 'rich_text' && prop.name && prop.name.toLowerCase().includes('author'))
      );
      
      if (authorProp) {
        if (authorProp.type === 'people' && authorProp.people && authorProp.people.length > 0) {
          return authorProp.people[0].name || 'Unknown';
        }
        if (authorProp.type === 'rich_text' && authorProp.rich_text && authorProp.rich_text.length > 0) {
          return authorProp.rich_text.map(t => t.plain_text).join('');
        }
      }
    }
    
    // Fallback to created_by
    if (page.created_by && page.created_by.name) {
      return page.created_by.name;
    }
    
    return 'Unknown Author';
  } catch (error) {
    console.error('Error extracting author:', error);
    return 'Unknown Author';
  }
}

// Check if page is public
function isPublicPage(page) {
  try {
    if (page.public_url) return true;
    
    // Check if page has public sharing enabled
    if (page.properties) {
      const statusProp = Object.values(page.properties).find(
        prop => prop.type === 'status' || 
               (prop.type === 'select' && prop.name && prop.name.toLowerCase().includes('status'))
      );
      
      if (statusProp) {
        if (statusProp.type === 'status' && statusProp.status) {
          return statusProp.status.name.toLowerCase() === 'public' ||
                 statusProp.status.name.toLowerCase() === 'published';
        }
        if (statusProp.type === 'select' && statusProp.select) {
          return statusProp.select.name.toLowerCase() === 'public' ||
                 statusProp.select.name.toLowerCase() === 'published';
        }
      }
    }
    
    // Default to considering it public if no clear indication
    return true;
  } catch (error) {
    console.error('Error checking public status:', error);
    return false;
  }
}

// Post message to Slack
async function postToSlack(page, title, author, url) {
  const message = {
    channel: SLACK_CHANNEL,
    blocks: [
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text: `📄 *New article published!*`
        }
      },
      {
        type: 'section',
        fields: [
          {
            type: 'mrkdwn',
            text: `*📝 Article:*\n<${url}|${title}>`
          },
          {
            type: 'mrkdwn',
            text: `*✍️ Author:*\n${author}`
          }
        ]
      },
      {
        type: 'divider'
      }
    ]
  };
  
  try {
    await slack.chat.postMessage(message);
    console.log(`✅ Posted to Slack: ${title}`);
  } catch (error) {
    console.error(`❌ Error posting to Slack:`, error.message);
    throw error;
  }
}

// Main function
async function main() {
  console.log('🚀 Starting Notion to Slack sync...');
  
  // Load cache
  const cache = loadCache();
  console.log(`📦 Loaded cache with ${cache.postedPages.length} posted pages`);
  
  // Get all pages from Notion root page
  console.log('🔍 Fetching pages from Notion root page...');
  const allPages = await getAllPagesRecursive(ROOT_PAGE_ID);
  console.log(`📄 Found ${allPages.length} total pages (excluding root)`);
  
  // Current date
  const oneWeekAgo = new Date();
  oneWeekAgo.setDate(oneWeekAgo.getDate() - 7);
  
  // Filter pages: created more than a week ago, public, and not already posted
  const pagesToPost = [];
  
  for (const page of allPages) {
    const pageId = page.id;
    const createdTime = new Date(page.created_time);
    const isOldEnough = createdTime < oneWeekAgo;
    const isPublic = isPublicPage(page);
    const alreadyPosted = cache.postedPages.includes(pageId);
    
    if (isOldEnough && isPublic && !alreadyPosted) {
      pagesToPost.push(page);
    }
  }
  
  console.log(`📬 Found ${pagesToPost.length} new pages to post`);
  
  // Post to Slack
  for (const page of pagesToPost) {
    const title = getPageTitle(page);
    const author = getAuthor(page);
    const url = page.url;
    
    console.log(`\n📤 Posting: "${title}" by ${author}`);
    
    try {
      await postToSlack(page, title, author, url);
      cache.postedPages.push(page.id);
    } catch (error) {
      console.error(`Failed to post page: ${title}`);
    }
    
    // Small delay to avoid rate limits
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  
  // Save updated cache
  saveCache(cache);
  console.log('\n✨ Sync complete!');
}

// Run the script
main().catch(error => {
  console.error('❌ Fatal error:', error);
  process.exit(1);
});

File: .github/scripts/notion_slack_sync.py
import os
import json
import requests
from datetime import datetime, timedelta
from pathlib import Path
from notion_client import Client

# Initialize Notion client
notion = Client(auth=os.environ.get('NOTION_TOKEN'))

# Configuration from environment variables
ROOT_PAGE_ID = os.environ.get('ROOT_PAGE_ID')
SLACK_CHANNEL = os.environ.get('SLACK_CHANNEL')
SLACK_WEBHOOK_URL = os.environ.get('SLACK_WEBHOOK_URL')

# Cache file configuration
CACHE_DIR = Path(__file__).parent.parent / 'cache'
CACHE_FILE = CACHE_DIR / 'posted_pages.json'

# Ensure cache directory exists
CACHE_DIR.mkdir(parents=True, exist_ok=True)


def load_cache():
    """Load cache of already posted pages"""
    try:
        if CACHE_FILE.exists():
            with open(CACHE_FILE, 'r') as f:
                return json.load(f)
    except Exception as e:
        print(f'Error loading cache: {e}')
    return {'posted_pages': []}


def save_cache(cache):
    """Save cache to file"""
    try:
        with open(CACHE_FILE, 'w') as f:
            json.dump(cache, f, indent=2)
    except Exception as e:
        print(f'Error saving cache: {e}')


def get_all_pages_recursive(page_id, pages=None, is_root=True):
    """Recursively get all pages from a root page"""
    if pages is None:
        pages = []
    
    try:
        # Get the page details (skip for root page as we don't want to include it)
        if not is_root:
            page = notion.pages.retrieve(page_id=page_id)
            if page.get('object') == 'page':
                pages.append(page)
        
        # Get all child blocks
        has_more = True
        start_cursor = None
        
        while has_more:
            response = notion.blocks.children.list(
                block_id=page_id,
                page_size=100,
                start_cursor=start_cursor
            )
            
            # Process each block
            for block in response.get('results', []):
                if block.get('type') == 'child_page':
                    # Recursively get this child page and its children
                    get_all_pages_recursive(block['id'], pages, False)
            
            has_more = response.get('has_more', False)
            start_cursor = response.get('next_cursor')
    
    except Exception as e:
        print(f'Error fetching pages for {page_id}: {e}')
    
    return pages


def get_page_title(page):
    """Extract page title"""
    try:
        if 'properties' in page:
            # Try different property types
            for prop_name, prop_value in page['properties'].items():
                if prop_value.get('type') == 'title':
                    title_content = prop_value.get('title', [])
                    if title_content:
                        return ''.join([t.get('plain_text', '') for t in title_content])
        
        return 'Untitled'
    except Exception as e:
        print(f'Error extracting title: {e}')
        return 'Untitled'


def get_author(page):
    """Get author from page"""
    try:
        # Try to find author in properties
        if 'properties' in page:
            for prop_name, prop_value in page['properties'].items():
                prop_type = prop_value.get('type')
                
                # Check for people property
                if prop_type == 'people':
                    people = prop_value.get('people', [])
                    if people:
                        return people[0].get('name', 'Unknown')
                
                # Check for rich_text property with 'author' in name
                if prop_type == 'rich_text' and 'author' in prop_name.lower():
                    rich_text = prop_value.get('rich_text', [])
                    if rich_text:
                        return ''.join([t.get('plain_text', '') for t in rich_text])
        
        # Fallback to created_by
        if 'created_by' in page:
            return page['created_by'].get('name', 'Unknown Author')
        
        return 'Unknown Author'
    except Exception as e:
        print(f'Error extracting author: {e}')
        return 'Unknown Author'


def is_public_page(page):
    """Check if page is public"""
    try:
        # Check if page has public URL
        if page.get('public_url'):
            return True
        
        # Check if page has public sharing enabled via properties
        if 'properties' in page:
            for prop_name, prop_value in page['properties'].items():
                prop_type = prop_value.get('type')
                
                # Check status property
                if prop_type == 'status':
                    status = prop_value.get('status', {})
                    status_name = status.get('name', '').lower()
                    if status_name in ['public', 'published']:
                        return True
                
                # Check select property with 'status' in name
                if prop_type == 'select' and 'status' in prop_name.lower():
                    select = prop_value.get('select', {})
                    if select:
                        select_name = select.get('name', '').lower()
                        if select_name in ['public', 'published']:
                            return True
        
        # Default to considering it public if no clear indication
        return True
    except Exception as e:
        print(f'Error checking public status: {e}')
        return False


def post_to_slack(page, title, author, url):
    """Post message to Slack using webhook"""
    message = {
        "channel": SLACK_CHANNEL,
        "blocks": [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": "📄 *New article published!*"
                }
            },
            {
                "type": "section",
                "fields": [
                    {
                        "type": "mrkdwn",
                        "text": f"*📝 Article:*\n<{url}|{title}>"
                    },
                    {
                        "type": "mrkdwn",
                        "text": f"*✍️ Author:*\n{author}"
                    }
                ]
            },
            {
                "type": "divider"
            }
        ]
    }
    
    try:
        response = requests.post(
            SLACK_WEBHOOK_URL,
            json=message,
            headers={'Content-Type': 'application/json'}
        )
        response.raise_for_status()
        print(f'✅ Posted to Slack: {title}')
    except Exception as e:
        print(f'❌ Error posting to Slack: {e}')
        raise


def main():
    print('🚀 Starting Notion to Slack sync...')
    
    # Load cache
    cache = load_cache()
    print(f"📦 Loaded cache with {len(cache['posted_pages'])} posted pages")
    
    # Get all pages from Notion root page
    print('🔍 Fetching pages from Notion root page...')
    all_pages = get_all_pages_recursive(ROOT_PAGE_ID)
    print(f'📄 Found {len(all_pages)} total pages (excluding root)')
    
    # Calculate date threshold (7 days ago)
    one_week_ago = datetime.now() - timedelta(days=7)
    
    # Filter pages: created more than a week ago, public, and not already posted
    pages_to_post = []
    
    for page in all_pages:
        page_id = page['id']
        created_time = datetime.fromisoformat(page['created_time'].replace('Z', '+00:00'))
        is_old_enough = created_time < one_week_ago
        is_public = is_public_page(page)
        already_posted = page_id in cache['posted_pages']
        
        if is_old_enough and is_public and not already_posted:
            pages_to_post.append(page)
    
    print(f'📬 Found {len(pages_to_post)} new pages to post')
    
    # Post to Slack
    for page in pages_to_post:
        title = get_page_title(page)
        author = get_author(page)
        url = page['url']
        
        print(f'\n📤 Posting: "{title}" by {author}')
        
        try:
            post_to_slack(page, title, author, url)
            cache['posted_pages'].append(page['id'])
        except Exception as e:
            print(f'Failed to post page: {title}')
        
        # Small delay to avoid rate limits
        import time
        time.sleep(1)
    
    # Save updated cache
    save_cache(cache)
    print('\n✨ Sync complete!')


if __name__ == '__main__':
    main()
